src/Features/LanguageServer/Protocol/Handler/Diagnostics/WorkspacePullDiagnosticHandler.cs

```diff
@@ -55,16 +55,17 @@ protected override DiagnosticTag[] ConvertTags(DiagnosticData diagnosticData)
 
         protected override ValueTask<ImmutableArray<Document>> GetOrderedDocumentsAsync(RequestContext context, CancellationToken cancellationToken)
         {
-            return GetWorkspacePullDocumentsAsync(context, DiagnosticService.GlobalOptions, cancellationToken);
+            return GetWorkspacePullDocumentsAsync(context, GlobalOptions, cancellationToken);
         }
 
         protected override Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(
             RequestContext context, Document document, DiagnosticMode diagnosticMode, CancellationToken cancellationToken)
         {
-            // For closed files, go to the IDiagnosticService for results.  These won't necessarily be totally up to
-            // date.  However, that's fine as these are closed files and won't be in the process of being edited.  So
-            // any deviations in the spans of diagnostics shouldn't be impactful for the user.
-            return DiagnosticService.GetPullDiagnosticsAsync(document, includeSuppressedDiagnostics: false, diagnosticMode, cancellationToken).AsTask();
+            // Directly use the IDiagnosticAnalyzerService.  This will use the actual snapshots
+            // we're passing in.  If information is already cached for that snapshot, it will be returned.  Otherwise,
+            // it will be computed on demand.  Because it is always accurate as per this snapshot, all spans are correct
+            // and do not need to be adjusted.
+            return DiagnosticAnalyzerService.GetDiagnosticsForSpanAsync(document, range: null, cancellationToken: cancellationToken);
```

Comment:
Was also thinking of switching this to this API - https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/IDiagnosticAnalyzerService.cs,61 which appears to use the solution passed in for diagnostics.  

The current implementation is technically missing diagnostics at the project level, but I think the majority of those are build diagnostics (which are published via CPS directly when LSP pull is on).  @mavasani do you know a non-build project level diagnostic I could test off the top of your head?

Reply:
> do you know a non-build project level diagnostic I could test off the top of your head

You would need to turn on full solution analysis for this - one possibility is testing an analyzer diagnostic from compilation end analyzers. These are only reported in live analysis when FSA is on, and they are treated as non-local project level diagnostics: https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Diagnostics/DiagnosticAnalysisResult.cs,33. You can enable [CA1812](https://github.com/dotnet/roslyn-analyzers/blob/e52a3f1c0b45684ddaa041f26a980fa451ea9001/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs#L19) which ships as a disabled analyzer in the .NET SDK via an editorconfig entry `dotnet_diagnostic.CA1812.severity = warning` and turning on full solution analysis.